app/browser/api/windowApi.js

```diff
@@ -0,0 +1,281 @@
+
+    browserOpts.x = Math.min(browserOpts.x, maxX - defaults.windowOffset)
+    browserOpts.y = Math.min(browserOpts.y, maxY - defaults.windowOffset)
+  }
+
+  const minWidth = isModal(browserOpts) ? defaults.minModalWidth : defaults.minWidth
+  const minHeight = isModal(browserOpts) ? defaults.minModalHeight : defaults.minHeight
+
+  // min width and height don't seem to work when the window is first created
+  browserOpts.width = browserOpts.width < minWidth ? minWidth : browserOpts.width
+  browserOpts.height = browserOpts.height < minHeight ? minHeight : browserOpts.height
+
+  const autoHideMenuBarSetting = isDarwin || getSetting(settings.AUTO_HIDE_MENU)
+
+  const windowProps = {
+    // smaller min size for "modal" windows
+    minWidth,
+    minHeight,
+    // Neither a frame nor a titlebar
+    // frame: false,
+    // A frame but no title bar and windows buttons in titlebar 10.10 OSX and up only?
+    titleBarStyle: 'hidden-inset',
+    autoHideMenuBar: autoHideMenuBarSetting,
+    title: appConfig.name,
+    webPreferences: defaults.webPreferences,
+    frame: !isWindows
+  }
+
+  if (process.platform === 'linux') {
+    windowProps.icon = path.join(__dirname, '..', '..', 'res', 'app.png')
+  }
+
+  const homepageSetting = getSetting(settings.HOMEPAGE)
+  const startupSetting = getSetting(settings.STARTUP_MODE)
+  const toolbarUserInterfaceScale = getSetting(settings.TOOLBAR_UI_SCALE)
+
+  setImmediate(() => {
+    let mainWindow = new BrowserWindow(Object.assign(windowProps, browserOpts, {disposition: frameOpts.disposition}))
+    let restoredImmutableWindowState = action.restoredState
+    initWindowCacheState(mainWindow.id, restoredImmutableWindowState)
+
+    // initialize frames state
+    let frames = Immutable.List()
+    if (restoredImmutableWindowState) {
+      frames = restoredImmutableWindowState.get('frames')
+      restoredImmutableWindowState = restoredImmutableWindowState.set('frames', Immutable.List())
+      restoredImmutableWindowState = restoredImmutableWindowState.set('tabs', Immutable.List())
+    } else {
+      if (frameOpts && Object.keys(frameOpts).length > 0) {
+        if (frameOpts.forEach) {
+          frames = Immutable.fromJS(frameOpts)
+        } else {
+          frames = frames.push(Immutable.fromJS(frameOpts))
+        }
+      } else if (startupSetting === 'homePage' && homepageSetting) {
+        frames = Immutable.fromJS(homepageSetting.split('|').map((homepage) => {
+          return {
+            location: homepage
+          }
+        }))
+      }
+    }
+
+    if (frames.size === 0) {
+      frames = Immutable.fromJS([{}])
+    }
+
+    if (immutableWindowState.getIn(['ui', 'isMaximized'])) {
+      mainWindow.maximize()
+    }
+
+    if (immutableWindowState.getIn(['ui', 'isFullScreen'])) {
+      mainWindow.setFullScreen(true)
+    }
+
+    mainWindow.webContents.on('did-finish-load', (e) => {
+      // lastEmittedState = state TODO what to do here?
```

Comment:
what should we do for this line. `lastEmittedState` is local variable in `appStore`